### PR TITLE
pgctl debug and pre-start/post-stop fixes

### PR DESCRIPTION
**Pull request by bchess**
A couple fixes for interaction between pre-start/post-stop and pgctl debug:

1. Do not run post-stop if a service is being shut down to enter debug mode.
2. Run pre-start prior to entering debug mode on a stopped service.

Internal ticket COREBACK-1796

### CODE REVIEW

File: pgctl/cli.py [hunk 1]
```diff
@@ -238,6 +239,10 @@ def __change_state(self, state):
                     for service in self.all_services
                 )
 
+        if self._entering_debug:
+            # Do not post-stop when we are about to start a debug session
+            run_post_stop_hook = False
+
```
Comment (asottile) [inline pgctl/cli.py]: I don't agree with this change.  debug should be equivalent to `pgctl stop && pgctl start in foreground`

Comment (asottile) [inline pgctl/cli.py]: the rest of the change is agreeable however

File: pgctl/cli.py [hunk 2]
```diff
@@ -410,8 +415,16 @@ def debug(self):
...
 
-        self.stop()
-        service.foreground()  # never returns
+        try:
+            self._entering_debug = True
+            if service.state['state'] != 'down':
+                self.stop()
+            else:
+                # If we were not previously running, then also run pre-start
+                self._run_playground_wide_hook('pre-start')
```
Comment (0xngold) [inline pgctl/cli.py]: This makes debug special with respect to pre-start. In the service already running case, the service will be stopped, and then started in foreground mode without pre-start running. It seems like pre-start should run in that case as well.

File: pgctl/cli.py (hunk 1 above)
Comment (0xngold) [inline pgctl/cli.py]: +1 to this. This won't break things for the single service case we're looking at internally, and it will be less surprising.

Comment (bchess) [inline pgctl/cli.py]: The reason the current behavior is weird is that `pgctl debug` will result in post-stop being called if its the only service in the playground, whereas it won't call post-stop if there's another service.

IMHO the manner that post-stop is called should be the same whether the playground has 1 or >1 services.

Currently pre-start is called upon every pgctl start, regardless of how many services are already running. They're not consistent

File: pgctl/cli.py [hunk 3]
```diff
@@ -410,8 +415,16 @@ def debug(self):
...
+            service.foreground()  # never returns
```
Comment (bchess) [inline pgctl/cli.py]: finally would only be called if there was an exception

File: pgctl/cli.py (hunk 1 above)
Comment (asottile) [inline pgctl/cli.py]: Then address that separately (making pre-start and post-stop consistent).  It's orthogonal to this review whose goal should be to call pre-start for the debug case.

Comment (bchess) [inline pgctl/cli.py]: I think ycp would still be ok if we nixed this actually. As long as pre-start is still called before debug actually begins.

I'm fine getting rid of this

**Comment by bchess** [comment]
Ok - leaving post-stop as-is. But we always call pre-start before debug.

**Comment by bchess** [comment]
@asottile you good with this now?

File: pgctl/cli.py [hunk 4]
```diff
@@ -410,7 +410,10 @@ def debug(self):
                 'Must debug exactly one service, not: ' + commafy(self.service_names),
             )
 
-        self.stop()
+        if service.state['state'] != 'down':
+            self.stop()
```
Comment (asottile) [inline pgctl/cli.py]: stop is idempotent, why this change?

File: tests/spec/examples.py [hunk 5]
```diff
@@ -795,6 +797,16 @@ def it_runs_before_starting_a_service(self):
...
 
+    @pytest.mark.usefixtures('in_example_dir')
+    def it_runs_before_debugging_a_service(self):
+        proc = Popen(('setsid', 'pgctl', 'debug', 'sweet'), stdin=PIPE, stdout=PIPE)
+        proc.stdin.close()
+        try:
+            assert proc.stdout.readline() == 'hello, i am a pre-start script in stdout\n'
+        finally:
+            ctrl_c(proc)
+            proc.wait()
```
Comment (asottile) [inline tests/spec/examples.py]: could probably use one of the services which exits immediately so  you don't need to do the setsid ctrl_c dance

File: pgctl/cli.py (hunk 4 above)
Comment (bchess) [inline pgctl/cli.py]: its idempotent but not a no-op. It emits an unnecessary and confusing "<service> is already stopped" message. It's already down, why bother?

File: tests/spec/examples.py (hunk 5 above)
Comment (bchess) [inline tests/spec/examples.py]: This is the only one that has a pre-start script. Adding it would require additional changes to their tests because output will change.

Comment (asottile) [inline tests/spec/examples.py]: It's easy to add another fixture

File: pgctl/cli.py (hunk 4 above)
Comment (asottile) [inline pgctl/cli.py]: This change is unrelated to the goal of the PR -- make pre-start run for debug.

I'd rather we address this bit in a separate PR so this can ship.  If we're deeming the "Service is already stopped" message as useless perhaps we should remove it entirely.

File: tests/spec/examples.py (hunk 5 above)
Comment (bchess) [inline tests/spec/examples.py]: Yep, that is another approach, but I'm not seeing how your suggestion is any better than this one. Is this a blocking concern for you?

File: pgctl/cli.py (hunk 4 above)
Comment (bchess) [inline pgctl/cli.py]: It's not useless if you're issuing a stop directly. It's confusing and unnecessary if you're attempting to debug a stopped service.

Comment (asottile) [inline pgctl/cli.py]: Are the "Stopping ..." "Stopped ..." messages not equally confusing and useless when debugging?

File: tests/spec/examples.py (hunk 5 above)
Comment (asottile) [inline tests/spec/examples.py]: No neither of these are blocking.  I intentionally didn't check the "Request changes" option -- was merely making suggestions.
